README.md
# Go forth and build! 🚀

## Tutorial: How to make your first commit 📹:

This video example shows how to:

- Edit your container code
- Raise a new pull request
- Deploy the newest version automatically 🚀

https://user-images.githubusercontent.com/1718624/216366290-9bcac918-8605-4cf5-a98a-75bb8f722966.mp4



By now, you have deployed a container, and in moments, you can visit your app
live!

> Be patient! In ~3 mins your app is live, at your [app url](https://container-dnvfeko.containers.anotherwebservice.com/). It even has a free SSL/TLS certificate 🔒 you're welcome!

You probably want to add code to your app. Good news, your app is ready right now to start coding, which is simple:

1. Edit your code
2. Commit your code
3. Push your code `git push origin main`

Your app will be automatically re-deployed with the latest code at: https://container-dnvfeko.containers.anotherwebservice.com/

> You app is deployed already and is working software. Gone are the days of spending weeks coding and then *another* week going to production. No. Go to production *early* and respond to change.

# Getting Started 💻 (locally on your laptop)

> Step 0: You need to download your repo to your computer:

```
git clone git@github.com:plsgivemeachane/container-dnvfeko.git
cd container-dnvfeko
```

> See an error? You might need to setup permissions [here's a guide how to setup repo clone permissions](https://docs.github.com/en/authentication/connecting-to-github-with-ssh/adding-a-new-ssh-key-to-your-github-account)

1. [Install docker](https://docs.docker.com/get-docker/)

2. Start your container locally: `docker-compose up`
3. Visit your app locally: http://127.0.0.1:5000/

## View your app locally

Visit: http://127.0.0.1:5000/

### Rebuild container (locally)
If you make changes to `Dockerfile`, then you need to rebuild your container image. To rebuild the container image:
```
docker-compose build
# or 
docker-compose up --build
```

# Start coding! Which framework did you choose?

Need some help to get started?

- [**Flask** quickstart guide](https://flask.palletsprojects.com/en/2.2.x/quickstart/) ⚗️ 🐍
- [**Django** quide](https://docs.djangoproject.com/en/4.1/topics/http/views/) 📰
- [**Ruby** quickstart guide](https://github.com/KarmaComputing/rails-quickstart) 💎

# Debugging

How do I turn on the debugger?

Enable a breakpoint by adding `breakpoint()` to your code, start your application and run to that point then in a terminal type:

```
docker attach container-dnvfeko
```
Ta-da! You'll be inside the [Python debugger](https://docs.python.org/3/library/pdb.html#module-pdb) ( ⬅️ Read this!)



## Questions

- How was this built? [All code is here](https://github.com/KarmaComputing/container-hosting)


CHANGELOG.md
# v0.0.16 (Thu Dec 07 2023)

#### ⚠️ Pushed to `main`

- Update index.js ([@plsgivemeachane](https://github.com/plsgivemeachane))

#### Authors: 1

- [@plsgivemeachane](https://github.com/plsgivemeachane)

---

# v0.0.15 (Thu Dec 07 2023)

#### ⚠️ Pushed to `main`

- Add files via upload ([@plsgivemeachane](https://github.com/plsgivemeachane))

#### Authors: 1

- [@plsgivemeachane](https://github.com/plsgivemeachane)

---

# v0.0.14 (Sun Dec 03 2023)

#### ⚠️ Pushed to `main`

- Update docker-compose.yml ([@plsgivemeachane](https://github.com/plsgivemeachane))

#### Authors: 1

- [@plsgivemeachane](https://github.com/plsgivemeachane)

---

# v0.0.13 (Sat Dec 02 2023)

#### ⚠️ Pushed to `main`

- Update entry.sh ([@plsgivemeachane](https://github.com/plsgivemeachane))

#### Authors: 1

- [@plsgivemeachane](https://github.com/plsgivemeachane)

---

# v0.0.12 (Sat Dec 02 2023)

#### ⚠️ Pushed to `main`

- Update entry.sh ([@plsgivemeachane](https://github.com/plsgivemeachane))

#### Authors: 1

- [@plsgivemeachane](https://github.com/plsgivemeachane)

---

# v0.0.11 (Sat Dec 02 2023)

#### ⚠️ Pushed to `main`

- Update entry.sh ([@plsgivemeachane](https://github.com/plsgivemeachane))

#### Authors: 1

- [@plsgivemeachane](https://github.com/plsgivemeachane)

---

# v0.0.10 (Sat Dec 02 2023)

#### ⚠️ Pushed to `main`

- Update entry.sh ([@plsgivemeachane](https://github.com/plsgivemeachane))

#### Authors: 1

- [@plsgivemeachane](https://github.com/plsgivemeachane)

---

# v0.0.9 (Sat Dec 02 2023)

#### ⚠️ Pushed to `main`

- Update entry.sh ([@plsgivemeachane](https://github.com/plsgivemeachane))

#### Authors: 1

- [@plsgivemeachane](https://github.com/plsgivemeachane)

---

# v0.0.8 (Sat Dec 02 2023)

#### ⚠️ Pushed to `main`

- Update Dockerfile ([@plsgivemeachane](https://github.com/plsgivemeachane))

#### Authors: 1

- [@plsgivemeachane](https://github.com/plsgivemeachane)

---

# v0.0.7 (Sat Dec 02 2023)

#### ⚠️ Pushed to `main`

- Update Dockerfile ([@plsgivemeachane](https://github.com/plsgivemeachane))

#### Authors: 1

- [@plsgivemeachane](https://github.com/plsgivemeachane)

---

# v0.0.6 (Sat Dec 02 2023)

#### ⚠️ Pushed to `main`

- Update entry.sh ([@plsgivemeachane](https://github.com/plsgivemeachane))

#### Authors: 1

- [@plsgivemeachane](https://github.com/plsgivemeachane)

---

# v0.0.5 (Sat Dec 02 2023)

#### ⚠️ Pushed to `main`

- Update Dockerfile ([@plsgivemeachane](https://github.com/plsgivemeachane))
- Create entry.sh ([@plsgivemeachane](https://github.com/plsgivemeachane))

#### Authors: 1

- [@plsgivemeachane](https://github.com/plsgivemeachane)

---

# v0.0.4 (Sat Dec 02 2023)

#### ⚠️ Pushed to `main`

- Dockerfile with HTTP ([@plsgivemeachane](https://github.com/plsgivemeachane))

#### Authors: 1

- [@plsgivemeachane](https://github.com/plsgivemeachane)

---

# v0.0.3 (Sat Dec 02 2023)

#### ⚠️ Pushed to `main`

- Update Dockerfile ([@plsgivemeachane](https://github.com/plsgivemeachane))

#### Authors: 1

- [@plsgivemeachane](https://github.com/plsgivemeachane)

---

# v0.0.2 (Sat Dec 02 2023)

#### ⚠️ Pushed to `main`

- Create Dockerfile ([@plsgivemeachane](https://github.com/plsgivemeachane))

#### Authors: 1

- [@plsgivemeachane](https://github.com/plsgivemeachane)

---

# v0.0.1 (Sat Dec 02 2023)

:tada: This release contains work from a new contributor! :tada:

Thank you, null[@plsgivemeachane](https://github.com/plsgivemeachane), for all your work!

#### ⚠️ Pushed to `main`

- Added deploy.yml file ([@plsgivemeachane](https://github.com/plsgivemeachane))
- Added release.yml file ([@plsgivemeachane](https://github.com/plsgivemeachane))
- Added deploy.sh file ([@plsgivemeachane](https://github.com/plsgivemeachane))
- Added .autorc file ([@plsgivemeachane](https://github.com/plsgivemeachane))
- Added amber.yaml secrets file ([@plsgivemeachane](https://github.com/plsgivemeachane))
- create .docker-compose.yml ([@plsgivemeachane](https://github.com/plsgivemeachane))
- create README.md ([@plsgivemeachane](https://github.com/plsgivemeachane))

#### Authors: 1

- [@plsgivemeachane](https://github.com/plsgivemeachane)


index.js
const CharacterAI = require('./cai/index.js');
const characterAI = new CharacterAI();
const express = require("express")
const app = express()
const cors = require("cors")
let doneSetup = false;
let chat;
let lock;

const sleep = ms => new Promise(r => setTimeout(r, ms));

function compress(string) {
  string = unescape(encodeURIComponent(string));
  var newString = '',
    char, nextChar, combinedCharCode;
  for (var i = 0; i < string.length; i += 2) {
    char = string.charCodeAt(i);

    if ((i + 1) < string.length) {


      nextChar = string.charCodeAt(i + 1) - 31;


      combinedCharCode = char + "" + nextChar.toLocaleString('en', {
        minimumIntegerDigits: 2
      });

      newString += String.fromCharCode(parseInt(combinedCharCode, 10));

    } else {


      newString += string.charAt(i);
    }
  }
  return btoa(unescape(encodeURIComponent(newString)));
}

async function init() {
  await characterAI.authenticateAsGuest()
  const characterId = "bQBlcCVVjfTAzv8OWK30dw7Tj-TNtKC_Rh0Z46Dx6fY" // Discord moderator
  chat = await characterAI.createOrContinueChat(characterId);
  console.log("Worked")
  doneSetup = true
}

async function bruh(prompt) {
  const response = await chat.generateImage(prompt, true)
  //console.log(response)
  return response
}

app.use(cors({
  origin: "*"
}))

app.get("/", async (req, res) => {
  while (!doneSetup) await sleep(1000)
  res.send("Ok")
})

app.get("/prompt/:pr", async (req, res) => {
  while (!doneSetup || lock) await sleep(1000)
  lock = true
  try {
    const url = await bruh(req.params.pr)
    res.send(url)
  }
  finally {
    lock = false
  }
})

app.listen(80, () => {
  console.log("Worked in 3000")
  init()
  lock = false;
})


cai/client.js
const Chat = require('./chat')
const { v4: uuidv4 } = require('uuid');
const Parser = require('./parser')
const Requester = require('./requester')

class Client {
    #token = undefined;
    #isGuest = false;
    #authenticated = false;
    #guestHeaders = {
        "content-type": "application/json",
        "user-agent": 'CharacterAI/1.0.0 (iPhone; iOS 14.4.2; Scale/3.00)'
    }
    requester = new Requester();

    constructor() {
        this.#token = undefined;
    }

    // api fetching
    async fetchCategories() {
        const request = await this.requester.request('https://beta.character.ai/chat/character/categories/')

        if (request.status() === 200) return await Parser.parseJSON(request);
        else throw Error('Failed to fetch categories');
    }
    async fetchUserConfig() {
        const request = await this.requester.request('https://beta.character.ai/chat/config/', {
            headers:this.#guestHeaders
        })
        
        if (request.status() === 200) {
            const response = await Parser.parseJSON(request)

            return response;
        } else Error('Failed fetching user configuration.')
    }
    async fetchUser() {
        if (!this.isAuthenticated()) throw Error('You must be authenticated to do this.');
        
        const request = await this.requester.request('https://beta.character.ai/chat/user/', {
            headers:this.getHeaders()
        })

        if (request.status() === 200) {
            const response = await Parser.parseJSON(request)

            return response;
        } else Error('Failed fetching user.')
    }
    async fetchFeaturedCharacters() {
        if (!this.isAuthenticated()) throw Error('You must be authenticated to do this.');
        
        const request = await this.requester.request('https://beta.character.ai/chat/characters/featured_v2/', {
            headers:this.getHeaders()
        })

        if (request.status() === 200) {
            const response = await Parser.parseJSON(request)

            return response;
        } else Error('Failed fetching featured characters.')
    }
    async fetchCharactersByCategory(curated = false) {
        if (curated == undefined || typeof(curated) != 'boolean') throw Error('Invalid arguments.')

        const url = `https://beta.character.ai/chat/${
            curated ? 'curated_categories' : 'categories'
        }/characters/`;

        const request = await this.requester.request(url, {
            headers:this.#guestHeaders
        })

        const property = curated
        ? 'characters_by_curated_category'
        : 'characters_by_category';
      
        if (request.status() === 200) {
            const response = await Parser.parseJSON(request)

            return response[property]
        } else Error('Failed fetching characters by category.')
    }
    async fetchCharacterInfo(characterId) {
        if (!this.isAuthenticated()) throw Error('You must be authenticated to do this.');
        if (characterId == undefined || typeof(characterId) != 'string') throw Error('Invalid arguments.')

        const request = await this.requester.request(`https://beta.character.ai/chat/character/info-cached/${characterId}/`, {
            headers:this.getHeaders(),
        })

        if (request.status() === 200) {
            const response = await Parser.parseJSON(request)

            return response.character;
        } else Error('Could not fetch character information.')
    }
    async searchCharacters(characterName) {
        if (!this.isAuthenticated()) throw Error('You must be authenticated to do this.');
        if (this.#isGuest) throw Error('Guest accounts cannot use the search feature.');
        if (characterName == undefined || typeof(characterName) != 'string') throw Error('Invalid arguments.')

        const request = await this.requester.request(`https://beta.character.ai/chat/characters/search/?query=${characterName}`, {
            headers:this.getHeaders()
        })
        
        if (request.status() === 200) {
            const response = await Parser.parseJSON(request)

            return response;
        } else Error('Could not search for characters.')
    }
    async getRecentConversations() {
        if (!this.isAuthenticated()) throw Error('You must be authenticated to do this.');
        const request = await this.requester.request(`https://beta.character.ai/chat/characters/recent/`, {
            headers:this.getHeaders()
        })

        if (request.status() === 200) {
            const response = await Parser.parseJSON(request)

            return response;
        } else Error('Could not get recent conversations.')
    }

    // chat
    async createOrContinueChat(characterId, externalId = null) {
        if (!this.isAuthenticated()) throw Error('You must be authenticated to do this.');
        if (characterId == undefined || typeof(characterId) != 'string' || typeof(externalId != null ? externalId : '') != 'string') throw Error('Invalid arguments.')

        let request = await this.requester.request('https://beta.character.ai/chat/history/continue/',  {
            body:Parser.stringify({
                character_external_id: characterId,
                history_external_id: externalId,
            }),
            method:'POST',
            headers:this.getHeaders()
        })

        if (request.status() === 200 || request.status() === 404) {
            let response = await request.text()
            
            if (response === "No Such History" || response === "there is no history between user and character") { // Create a new chat
                request = await this.requester.request('https://beta.character.ai/chat/history/create/', {
                    body:Parser.stringify({
                        character_external_id: characterId,
                        history_external_id: null,
                    }),
                    method:'POST',
                    headers:this.getHeaders()
                })
                if (request.status() === 200) response = await Parser.parseJSON(request)
                else Error('Could not create a new chat.')
            } 

            // If a text gets returned, we try to parse it to JSON!
            try {
                response = JSON.parse(response);
            } catch (error) {}

            // Continue it 
            const continueBody = response;
            return new Chat(this, characterId, continueBody)
        } else Error('Could not create or resume a chat.')
    }

    // authentification
    async authenticateWithToken(token) {
        if (this.isAuthenticated()) throw Error('Already authenticated');
        if (!token || typeof(token) != 'string') throw Error('Specify a valid token');
        await this.requester.initialize();

        console.log("We are here !")

        const request = await this.requester.request('https://beta.character.ai/dj-rest-auth/auth0/', {
            method:'POST',
            body:Parser.stringify({
                access_token: token
            }),
            headers:{
                'Content-Type': 'application/json',
            }
        })

        console.log(request)

        if (request.status() === 200) {
            const response = await Parser.parseJSON(request)

            this.#isGuest = false;
            this.#authenticated = true;
            this.#token = response.key;

            return response.token
        } else throw Error('Token is invalid')
    }
    async authenticateAsGuest() {
        if (this.isAuthenticated()) throw Error('Already authenticated');
        await this.requester.initialize();

        let generating = false;
        let request;

        let uuid = uuidv4();

        // This is experimental but forces authentication 
        for (let i = 0; i < 20; i++) {
            generating = true;
          
            uuid = uuidv4();
            const payload = JSON.stringify({
                lazy_uuid: uuid,
            });
          
            const baseRequest = await Promise.race([
                this.requester.request('https://beta.character.ai/chat/auth/lazy/', {
                    method: 'POST',
                    body: payload,
                    headers: this.#guestHeaders,
                }),
                new Promise(resolve => setTimeout(() => resolve(null), 2000))
            ]);
          
            request = baseRequest;
            generating = false;
          
            if (request) break;
        }

        if (request.status() === 200) {
            const response = await Parser.parseJSON(request)
            
            if (response.success === true) {
                this.#isGuest = true;
                this.#authenticated = true;
                this.#token = response.token;
                this.uuid = uuid;

                return response.token;
            } else throw Error('Registering failed');
        } else throw Error('Failed to fetch a lazy token')
    }
    unauthenticate() {
        if (this.isAuthenticated()) {
            this.#authenticated = false;
            this.#isGuest = false;
            this.#token = undefined;
            this.requester.uninitialize();
        }
    }

    // getters
    getToken() {
        return this.#token;
    }
    isGuest() {
        return this.#isGuest;
    }
    isAuthenticated() {
        return (this.#authenticated)
    }

    // headers
    getHeaders() {
        return {
            authorization: `Token ${this.#token}`,
            'Content-Type': 'application/json',
            //"user-agent": 'CharacterAI/1.0.0 (iPhone; iOS 14.4.2; Scale/3.00)'
            //'sec-ch-ua': `"Not_A Brand";v="99", "Google Chrome";v="109", "Chromium";v="109"`,
            //'sec-ch-ua-mobile': '?0'
            /*'sec-ch-ua-platform': "Windows",
            'sec-fetch-dest': 'empty',
            'sec-fetch-mode': 'cors',
            'sec-fetch-site': 'same-origin',*/
            //'user-agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/109.0.0.0 Safari/537.36'
        };
    }
}

module.exports = Client

cai/requester.js
const puppeteer = require('puppeteer-extra');
const StealthPlugin = require('puppeteer-extra-plugin-stealth')

class Requester {
    browser = undefined;
    page = undefined;

    #initialized = false;
    #hasDisplayed = false;

    #headless = "new";
    puppeteerPath = undefined;
    puppeteerLaunchArgs = [
        '--fast-start',
        '--disable-extensions',
        '--no-sandbox',
        '--disable-setuid-sandbox',
        '--no-gpu',
        '--disable-background-timer-throttling',
        '--disable-renderer-backgrounding',
        '--override-plugin-power-saver-for-testing=never',
        '--disable-extensions-http-throttling',
        '--user-agent=Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/109.0.0.0 Safari/537.3'
    ];
    puppeteerNoDefaultTimeout = false;
    puppeteerProtocolTimeout = 0;

    usePlus = false;
    forceWaitingRoom = true;

    constructor() {

    }
    isInitialized() {
        return this.#initialized;
    }

    async waitForWaitingRoom(page) {
        // Enable force waiting room to ensure you check for waiting room even on cai+.
        console.log("[node_characterai] Puppeteer - Enabling force waiting room...");
        if (!this.usePlus || (this.usePlus && this.forceWaitingRoom)) {
            return new Promise(async(resolve) => {
                try {
                    let interval;
                    let pass = true;
                    await page.goto("https://beta.character.ai");
                    
                    const minute = 1000; // Update every minute


                    // Keep waiting until false
                    async function check() {
                        if (pass) {
                            pass = false;
                            console.log("[node_characterai] Puppeteer - Checking for cloudflare's waiting room.");
                            const waitingRoomTimeLeft = await page.evaluate(async() => {
                                console.log("[node_characterai] Puppeteer - Waiting room time left");
                                try {
                                    console.log("[node_characterai] Puppeteer - Waiting room time left");
                                    const contentContainer = document.querySelector(".content-container");
                                    console.log("🚀 ~ file: requester.js:59 ~ Requester ~ waitingRoomTimeLeft ~ const contentContainer = document.querySelector(.content-container)");
                                    const sections = contentContainer.querySelectorAll("section");
                                    const h2Element = sections[i].querySelector("h2");
                                    const h2Text = h2Element.innerText;
                                    console.log(h2Text)
                                    const regex = /\d+/g;
                                    const matches = h2Text.match(regex);
                                    
                                    if (matches) return matches[0];
                                } catch (error) {
                                    console.log(error);
                                    return null;
                                };
                            }, minute);

                            console.log("[node_characterai] Puppeteer - Waiting room time left: " + waitingRoomTimeLeft);
                            
                            const waiting = (waitingRoomTimeLeft != null);
                            if (waiting) {
                                console.log(`[node_characterai] Puppeteer - Currently in cloudflare's waiting room. Time left: ${waitingRoomTimeLeft}`);
                            } else {
                                clearInterval(interval);
                                resolve();
                            }
                            pass = true;
                        };
                    }

                    interval = setInterval(check, minute);
                    await check();
                } catch (error) {
                    console.log(`[node_characterai] Puppeteer - There was a fatal error while checking for cloudflare's waiting room.`);
                    console.log(error);
                }
            });
        }
    }

    async initialize() {
        if (!this.isInitialized());
        
        // Handle chromium tabs cleanup
        process.on('exit', () => {
            this.uninitialize();
        });

        console.log("[node_characterai] Puppeteer - This is an experimental feature. Please report any issues on github.");

        puppeteer.use(StealthPlugin());
        const browser = await puppeteer.launch({
            headless: true,
            args: this.puppeteerLaunchArgs,
            protocolTimeout: this.puppeteerProtocolTimeout || 0, // Props to monckey100
            executablePath: '/nix/store/x205pbkd5xh5g4iv0g58xjla55has3cx-chromium-108.0.5359.94/bin/chromium-browser'
        });
        this.browser = browser;

        let page = await browser.pages();
        page = page[0];
        this.page = page;

        // Special thanks to @Parking-Master for this fix
        page.deleteCookie();
        
        const client = await page.target().createCDPSession();
        await client.send('Network.clearBrowserCookies');
        await client.send('Network.clearBrowserCache');

        await page.setRequestInterception(false);

        page.setViewport({
            width: 1920 + Math.floor(Math.random() * 100),
            height: 3000 + Math.floor(Math.random() * 100),
            deviceScaleFactor: 1,
            hasTouch: false,
            isLandscape: false,
            isMobile: false,
        });
        await page.setJavaScriptEnabled(true);
        await page.setDefaultNavigationTimeout(0);
        if (this.puppeteerNoDefaultTimeout) await page.setDefaultTimeout(0); // Props to monckey100

        const userAgent = 'CharacterAI/1.0.0 (iPhone; iOS 14.4.2; Scale/3.00)';
        await page.setUserAgent(userAgent);

        await this.waitForWaitingRoom(page);

        console.log("[node_characterai] Puppeteer - Done with setup");
    }

    /**
     * Asynchronously sends a request to the specified URL with the given options.
     *
     * @param {string} url - The URL to send the request to.
     * @param {Object} options - The options for the request.
     * @param {string} options.method - The HTTP method for the request.
     * @param {Object} options.body - The body of the request.
     * @param {Object} options.headers - The headers for the request.
     * @return {Promise<Response>} A Promise that resolves to the response from the request.
     */
    async request(url, options) {
        const page = this.page;

        const method = options.method;

        const body = (method == 'GET' ? {} : options.body);
        const headers = options.headers;

        let response

        if (this.usePlus){ // Props to @Kaidesa
            url = url.replace('beta.character.ai', 'plus.character.ai');
            console.log("Are we here?")
        }
        
        try {
            const payload = {
                method: method,
                headers: headers,
                body: body
            }

            if (url.endsWith("/streaming/")) {
                await page.setRequestInterception(false);
                if (!this.#hasDisplayed) {
                    console.log("[node_characterai] Puppeteer - Eval-fetching is an experimental feature and may be slower. Please report any issues on github")
                    this.#hasDisplayed = true;
                }

                // Bless @roogue & @drizzle-mizzle for the code here!
                response = await page.evaluate(
                    async (payload, url) => {
                        const response = await fetch(url, payload);

                        const data = await response.text();
                        const matches = data.match(/\{.*\}/g);

                        const responseText = matches[matches.length - 1];

                        let result = {
                            code: 500,
                        }

                        if (!matches) result = null;
                        else {
                            result.code = 200;
                            result.response = responseText;
                        }

                        return result;
                    },
                    payload,
                    url
                );

                response.status = () => response.code // compatibilty reasons
                response.text = () => response.response // compatibilty reasons
            } else {
                await page.setRequestInterception(true);
                let initialRequest = true;

                page.once('request', request => {
                    var data = {
                        'method': method,
                        'postData': body,
                        'headers': headers
                    };

                    if (request.isNavigationRequest() && !initialRequest) {
                        return request.abort();
                    }

                    try {
                        initialRequest = false;
                        request.continue(data);
                    } catch (error) {
                        console.log("[node_characterai] Puppeteer - Non fatal error: " + error);
                    }
                });
                response = await page.goto(url, { waitUntil: 'networkidle2' });
            }
        } catch (error) {
            const authenticating = (url == "https://beta.character.ai/chat/auth/lazy/")
            
            if (!authenticating) // Temporary fix
                console.log("[node_characterai] Puppeteer - " + error);
        }

        return response;
    }
    
    async uploadBuffer(buffer, client) {
        const page = this.page;

        let response

        try {
            await page.setRequestInterception(false);

            response = await page.evaluate(
                async (headers, buffer) => {
                        var result = {
                            code: 500
                        };

                        const blob = new Blob([buffer], { type: 'image/png' });
                        const formData = new FormData();
                        formData.append("image", blob, 'image.png');

                        let head = headers;
                        delete head["Content-Type"];
                        // ^ Is this even being used?

                        const uploadResponse = await fetch("https://beta.character.ai/chat/upload-image/", {
                            headers: headers,
                            method: "POST",
                            body: formData
                        })

                        if (uploadResponse.status == 200) {
                            result.code = 200;

                            let uploadResponseJSON = await uploadResponse.json();
                            result.response = uploadResponseJSON.value;
                        }

                        return result;
                    },
                    client.getHeaders(), buffer
            );

            response.status = () => response.code // compatibilty reasons
            response.body = () => response.response // compatibilty reasons
        } catch (error) {
            console.log("[node_characterai] Puppeteer - " + error)
        }

        return response;
    }

    async uninitialize() {
        // Handle chromium tabs cleanup
        try {
            this.browser.close();
        } catch {}
    }
}

module.exports = Requester;
